schema/encoding/jsonschema/benchmark_complex_test.go

```diff
@@ -0,0 +1,221 @@
+			"l": Optional(String(0, 65535, "l")),
+			"m": schema.Field{
+				Description: "m",
+				Validator: &schema.Array{
+					ValuesValidator: &schema.Object{
+						Schema: mSchema,
+					},
+				},
+			},
+		},
+	}
+	mSchema = &schema.Schema{
+		Description: "m",
+		Fields: schema.Fields{
+			"ma": String(0, 64, "ma"),
+			"mb": String(0, 64, "mb"),
+		},
+	}
+
+	responseSchema = &schema.Schema{
+		Description: "example response",
+		Fields: schema.Fields{
+			"a": String(16, 16, "a"),
+			"b": String(0, 56, "b"),
+			"c": schema.Field{
+				Description: "c",
+				Required:    true,
+				Validator: &schema.Array{
+					ValuesValidator: &schema.Object{
+						Schema: cSchema,
+					},
+				},
+			},
+			"d": schema.Field{
+				Description: "d",
+				Required:    true,
+				Validator: &schema.Array{
+					ValuesValidator: &schema.Object{
+						Schema: dSchema,
+					},
+				},
+			},
+			"e": String(0, 65335, "e"),
+			"f": schema.Field{
+				Validator:   &schema.String{},
+				Description: "f",
+			},
+			"g": schema.Field{
+				Validator:   &schema.String{},
+				Description: "g",
+			},
+		},
+	}
+	cSchema = &schema.Schema{
+		Description: "c schema",
+		Fields: schema.Fields{
+			"ca": rfc3339Nano(),
+			"cb": String(0, 64, "cb"),
+			"cc": Integer(0, 65535, "cc"),
+			"cd": Integer(0, 65535, "cd"),
+			"ce": Integer(0, math.MaxInt64, "ce"),
+			"cf": String(0, 128, "cf"),
+			"cg": String(0, 128, "cg"),
+			"ch": Integer(0, math.MaxInt64, "ch"),
+			"ci": Integer(0, math.MaxInt64, "ci"),
+			"cj": Integer(0, math.MaxInt64, "cj"),
+			"ck": Integer(0, math.MaxInt64, "ck"),
+		},
+	}
+	gSchema = &schema.Schema{
+		Description: "gSchema",
+		Fields: schema.Fields{
+			"ga": String(0, 64, "ga"),
+			"gb": String(0, 64, "gb"),
+			"gc": String(0, 64, "gc"),
+			"gd": String(0, 64, "gd"),
+			"ge": String(0, 64, "ge"),
+		},
+	}
+	iSchema = &schema.Schema{
+		Description: "iSchema",
+		Fields: schema.Fields{
+			"ia": Integer(0, math.MaxInt64, "ia"),
+			"ib": Integer(0, math.MaxInt64, "ib"),
+			"ic": Integer(0, math.MaxInt64, "ic"),
+			"id": rfc3339Nano(),
+		},
+	}
+	dSchema = &schema.Schema{
+		Description: "dSchema",
+		Fields: schema.Fields{
+			"da": String(0, 128, "da"),
+			"db": String(0, 32, "db"),
+			"dc": rfc3339Nano(),
+			"dd": rfc3339Nano(),
+			"de": Integer(0, 99999, "de"),
+			"df": Integer(0, 100, "df"),
+			"dg": schema.Field{
+				Description: "dg",
+				Validator: &schema.Object{
+					Schema: gSchema,
+				},
+			},
+			"dh": Bool("dh"),
+			"di": schema.Field{
+				Description: "di",
+				Validator: &schema.Object{
+					Schema: iSchema,
+				},
+			},
+		},
+	}
+)
+
+func BenchmarkEncoderComplex(b *testing.B) {
+	buf := new(bytes.Buffer)
+	b.Run("request", func(b *testing.B) {
+		for i := 0; i < b.N; i++ {
+			buf.Truncate(0)
+			enc := jsonschema.NewEncoder(buf)
+			enc.Encode(requestSchema)
+		}
+	})
+	b.Run("response", func(b *testing.B) {
+		for i := 0; i < b.N; i++ {
+			buf.Truncate(0)
```

Comment:
I understand but I'm more concerned we are measuring the noise of new, rather than the code under test. One thing specifically I want to see is the effect of all the buffer allocation vs just simply writing to a buffer. The point of the benchmark is to test the code, not new(bytes.Buffer) from the runtime. Turning this into a truncate amortizes the cost once across all the benchmarks and effectively removes it from the test.